libyara/modules/dex.c

```diff
@@ -0,0 +1,650 @@
+    printf("method(%i) = %s->%s(%s)%s\n", i, class_name, method_name, prototype, return_type);
+    */
+  }
+  yr_free(method_ids);
+}
+
+
+void load_map_list(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = *dex_header->map_offset;
+  uint8_t *pmap_list = data + offset;
+  size_t map_size = *pmap_list;
+  size_t map_data_size = sizeof(MAP_LIST) + (map_size * sizeof(MAP_ITEM));
+  if (offset + map_data_size > data_size) {
+    set_integer(1, module, "invalid_dex");
+    return;
+  }
+
+  MAP_LIST *map_list = yr_malloc(map_data_size);
+  memcpy(map_list, pmap_list, map_data_size);
+
+  set_integer(*map_list->size, module, "map_list.size");
+  int i = 0;
+  for (i = 0; i < map_size; i++) {
+    set_integer(*map_list->map_items[i].type, module, "map_list.map_items[%i].type", i);
+    set_integer(*map_list->map_items[i].size, module, "map_list.map_items[%i].size", i);
+    set_integer(*map_list->map_items[i].offset, module, "map_list.map_items[%i].offset", i);
+  }
+  yr_free(map_list);
+}
+
+
+char *get_string_item(uint32_t index, YR_OBJECT *module) {
+    SIZED_STRING *sized_string = get_string(module, "string_ids[%i].value", index);
+    char *string = yr_malloc(sized_string->length + 1);
+    strcpy(string, sized_string->c_string);
+
+    // c_string is supposed to be null terminated, but it's not for some reason
+    string[sized_string->length] = '\0';
+
+    return string;
+}
+
+
+char *get_prototype_string(uint16_t proto_idx, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = get_integer(module, "proto_ids[%i].parameters_offset", proto_idx);
+  if (offset == 0) {
+    return "";
+  }
+
+  uint8_t *ptype_list = data + offset;
+  size_t type_list_size = *ptype_list;
+  int size = sizeof(TYPE_LIST) + (type_list_size * sizeof(TYPE_ITEM));
+  if ((unsigned long)(size) > data_size) {
+    return "";
+  }
+
+  TYPE_LIST *type_list = yr_malloc(size);
+  memcpy(type_list, ptype_list, size);
+  char *params[type_list_size];
+  int string_size = 0;
+  int i = 0;
+  for (i = 0; i < type_list_size; i++) {
+    char *param = get_string_item(get_integer(module, "type_ids[%i].descriptor_idx", *type_list->type_items[i].type_idx), module);
+    string_size += strlen(param);
+    params[i] = param;
+  }
+
+  char *parameters = yr_malloc(string_size + 1);
+  strcpy(parameters, "");
+  for (i = 0; i < type_list_size; i++) {
+    strcat(parameters, params[i]);
+  }
+  parameters[string_size] = '\0';
+
+  return parameters;
+}
+
+
+uint64_t read_uleb128(uint8_t **buf, unsigned *uleb_size) {
+  uint8_t *ptr = *buf;
+  uint64_t value = 0;
+  unsigned shift = 0;
+  do {
+    value += ((uint64_t)(*ptr & 0x7f)) << shift;
+    shift += 7;
+  } while (*ptr++ >= 128);
+  if (uleb_size != NULL) {
+    *uleb_size = (unsigned)(ptr - *buf);
+  }
+  *buf = ptr;
+
+  return value;
+}
+
+
+uint64_t get_bytes_until_null(const uint8_t **buf) {
```

Comment:
this is done by strlen, when working with buffers is usually good to know the size of it to avoid oobs